```
--- FAIL: TestFlowLogStateUpgradeV0_complexState (0.00s)
panic: interface conversion: interface {} is nil, not *conns.AWSClient [recovered, repanicked]

goroutine 103 [running]:
testing.tRunner.func1.2({0x20a05940, 0x10b939e02bd0})
	/home/runner/sdk/go1.26.2/src/testing/testing.go:1974 +0x232
testing.tRunner.func1()
	/home/runner/sdk/go1.26.2/src/testing/testing.go:1977 +0x349
panic({0x20a05940?, 0x10b939e02bd0?})
	/home/runner/sdk/go1.26.2/src/runtime/panic.go:860 +0x13a
github.com/hashicorp/terraform-provider-aws/internal/service/ec2.flowLogStateUpgradeV0({0x23954000, 0x10b93b7bb720}, 0x2c0?, {0x0, 0x0})
	/home/runner/work/terraform-provider-aws/terraform-provider-aws/internal/service/ec2/vpc_flow_log_migrate.go:128 +0x1cc
github.com/hashicorp/terraform-provider-aws/internal/service/ec2_test.TestFlowLogStateUpgradeV0_complexState(0x10b93b7f2008)
	/home/runner/work/terraform-provider-aws/terraform-provider-aws/internal/service/ec2/vpc_flow_log_migrate_test.go:95 +0xd7
testing.tRunner(0x10b93b7f2008, 0x234a0648)
	/home/runner/sdk/go1.26.2/src/testing/testing.go:2036 +0xea
created by testing.(*T).Run in goroutine 1
	/home/runner/sdk/go1.26.2/src/testing/testing.go:2101 +0x4c5
FAIL	github.com/hashicorp/terraform-provider-aws/internal/service/ec2	1.790s
```

Maybe convert `TestFlowLogStateUpgradeV0_complexState` to an acceptance test?

--- a/internal/service/ec2/vpc_flow_log.go
+++ b/internal/service/ec2/vpc_flow_log.go
@@ -406,6 +406,10 @@ func flowLogARN(ctx context.Context, c *conns.AWSClient, flowLogID string) strin
 	return c.RegionalARN(ctx, names.EC2, "vpc-flow-log/"+flowLogID)
 }
 
-func cloudwatchLogGroupARNFromName(ctx context.Context, c *conns.AWSClient, logGroupName string) string {
+func cloudwatchLogGroupARNFromName(ctx context.Context, c regionalARNMaker, logGroupName string) string {
 	return c.RegionalARN(ctx, names.Logs, "log-group:"+logGroupName)
+}
+
+type regionalARNMaker interface {
+	RegionalARN(ctx context.Context, service, resource string) string
 }

--- a/internal/service/ec2/vpc_flow_log_migrate.go
+++ b/internal/service/ec2/vpc_flow_log_migrate.go
@@ -7,7 +7,6 @@ import (
 	"context"
 
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
-	"github.com/hashicorp/terraform-provider-aws/internal/conns"
 	tftags "github.com/hashicorp/terraform-provider-aws/internal/tags"
 	"github.com/hashicorp/terraform-provider-aws/names"
 )
@@ -125,7 +124,7 @@ func flowLogStateUpgradeV0(ctx context.Context, rawState map[string]any, meta an
 
 	name, ok := rawState[names.AttrLogGroupName]
 	if ok && name != "" {
-		c := meta.(*conns.AWSClient)
+		c := meta.(regionalARNMaker)
 
 		rawState["log_destination"] = cloudwatchLogGroupARNFromName(ctx, c, name.(string))
 	}

--- a/internal/service/ec2/vpc_flow_log_migrate_test.go
+++ b/internal/service/ec2/vpc_flow_log_migrate_test.go
@@ -10,6 +10,7 @@ import (
 
 	"github.com/aws/aws-sdk-go-v2/aws/arn"
 	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
 	tfec2 "github.com/hashicorp/terraform-provider-aws/internal/service/ec2"
 	"github.com/hashicorp/terraform-provider-aws/names"
 )

--- a/internal/service/ec2/vpc_flow_log_migrate_test.go
+++ b/internal/service/ec2/vpc_flow_log_migrate_test.go
@@ -4,9 +4,11 @@
 package ec2_test
 
 import (
+	"context"
 	"encoding/json"
 	"testing"
 
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
 	"github.com/google/go-cmp/cmp"
 	tfec2 "github.com/hashicorp/terraform-provider-aws/internal/service/ec2"
 	"github.com/hashicorp/terraform-provider-aws/names"
@@ -41,6 +43,7 @@ func TestFlowLogStateUpgradeV0(t *testing.T) {
 				names.AttrSubnetID:     "sn-12345678",
 			},
 			expected: map[string]any{
+				"log_destination":  "arn:aws:logs:us-east-1:123456789012:log-group:log-group-name",
 				names.AttrSubnetID: "sn-12345678",
 			},
 		},
@@ -50,7 +53,12 @@ func TestFlowLogStateUpgradeV0(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
 
-			result, err := tfec2.FlowLogStateUpgradeV0(t.Context(), tt.rawState, nil)
+			client := mockClient{
+				region:    "us-east-1",
+				accountID: acctest.Ct12Digit,
+			}
+
+			result, err := tfec2.FlowLogStateUpgradeV0(t.Context(), tt.rawState, client)
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}
@@ -92,7 +100,12 @@ func TestFlowLogStateUpgradeV0_complexState(t *testing.T) {
 		t.Fatalf("failed to unmarshal state JSON: %v", err)
 	}
 
-	result, err := tfec2.FlowLogStateUpgradeV0(t.Context(), rawState, nil)
+	client := mockClient{
+		region:    "us-east-1",
+		accountID: acctest.Ct12Digit,
+	}
+
+	result, err := tfec2.FlowLogStateUpgradeV0(t.Context(), rawState, client)
 	if err != nil {
 		t.Fatalf("unexpected error: %v", err)
 	}
@@ -104,4 +117,19 @@ func TestFlowLogStateUpgradeV0_complexState(t *testing.T) {
 	if result["log_destination"] != "arn:aws:logs:us-east-1:123456789012:log-group:/my/log-group" {
 		t.Errorf("expected log_destination to be preserved, got: %v", result["log_destination"])
 	}
+}
+
+type mockClient struct {
+	region    string
+	accountID string
+}
+
+func (m mockClient) RegionalARN(ctx context.Context, service, resource string) string {
+	return arn.ARN{
+		Partition: names.PartitionForRegion(m.region).ID(),
+		Service:   service,
+		Region:    m.region,
+		AccountID: m.accountID,
+		Resource:  resource,
+	}.String()
 }

--- a/internal/service/ec2/vpc_flow_log_test.go
+++ b/internal/service/ec2/vpc_flow_log_test.go
@@ -1608,9 +1608,9 @@ data "aws_partition" "current" {}
 func testAccVPCFlowLogConfig_upgradeFromV5LogGroupName_current(rName string) string {
 	return acctest.ConfigCompose(testAccFlowLogConfig_base(rName), fmt.Sprintf(`
 resource "aws_flow_log" "test" {
-  iam_role_arn   = aws_iam_role.test.arn
-  traffic_type   = "ALL"
-  vpc_id         = aws_vpc.test.id
+  iam_role_arn = aws_iam_role.test.arn
+  traffic_type = "ALL"
+  vpc_id       = aws_vpc.test.id
 
   log_destination      = aws_cloudwatch_log_group.test.arn
   log_destination_type = "cloud-watch-logs"

